src/frequenz/sdk/data_ingestion/resampling/timeseries_resampling_actor.py

```diff
@@ -0,0 +1,159 @@
+"""
+TimeSeriesResamplerActor used to subscribe for resampled component metrics.
+
+Copyright
+Copyright © 2022 Frequenz Energy-as-a-Service GmbH
+
+License
+MIT
+"""
+import logging
+from typing import Callable, Dict, Optional, Sequence
+
+from frequenz.channels import Broadcast, MergeNamed, Receiver, Select, Timer
+
+from frequenz.sdk.actor import actor
+from frequenz.sdk.core.timeseries import Sample, TimeSeriesId
+from frequenz.sdk.data_ingestion.resampling.data_provider import DataProvider
+from frequenz.sdk.data_ingestion.resampling.subscriptions import (
+    Subscription,
+    SubscriptionRequest,
+    SubscriptionResponse,
+)
+from frequenz.sdk.data_ingestion.resampling.timeseries_resampler import (
+    TimeSeriesResampler,
+)
+
+logger = logging.Logger(__name__)
+
+ResamplingFunction = Callable[[Sequence[Sample], float], float]
+
+
+# pylint: disable=unused-argument
+def average(samples: Sequence[Sample], resampling_frequency: float) -> float:
+    """Calculate average of the provided values.
+
+    Args:
+        samples: sequences of samples to apply the average to
+        resampling_frequency: resampling frequency, in seconds
+
+    Returns:
+        average of all the sample values
+    """
+    if len(samples) == 0:
+        return 0.0
```

Comment:
This should actually never happen. I think this should be an assertion (and we should put in the documentation that samples can't be empty.

Reply:
Why can't it happen? Consider this scenario:
- a consumer subscribes for a given `TimeSeriesId`
- resampling actor creates an input channel for that `TimeSeriesId` to receives data from `DataSourcingActor` and never receives any data through that channel (for whatever reason)
- resampling actor also also indirectly creates an instance of `InvividualTimeSeriesResampler` for that `TimeSeriesId`
- resampling timer triggers a resampling round, which at the end calls `InvividualTimeSeriesResampler.resample()`, which in turn calls `average(self._ring_buffer.items, self._period_s)` where ring buffer has no samples

I am not sure if `0.0` should be returned in this case, but for sure wanted to avoid division by zero.

Reply:
> Why can't it happen?

Because the caller should not call the function with an empty list :)

That's it, is a precondition, it doesn't make sense to calculate an average over no data.

Reply:
Ok, it makes sense that there's no point in calculating average of an empty list. 

I will change the `resample` method in the `IndividualTimeSeriesResampler` to return `None` if its ring buffer is empty so that it will looks like this:
```
class TimeSeriesResampler:
    def resample(self) -> Iterator[Tuple[TimeSeriesId, Optional[Sample]]]:
        now = datetime.now(tz=pytz.UTC)
        for time_series_id, resampler in self._resamplers.items():
            yield time_series_id, Sample(timestamp=now, value=resampler.resample())
```
calls
```
class IndividualTimeSeriesResampler:
    def resample(self) -> Optional[float]:
        if len(self._ring_buffer.items) <= 0:
               return None
        return self._resampling_function(self._ring_buffer.items, self._period_s)
```

Reply:
To me it makes sense to handle the case when the input list is empty. There is no guarantee how a user is going to use it, and regardless of how much documentation we put in, sometime somehow it will get an empty list.

I'd suggest raising an exception in this case, and specifying in docs when the exception is raised.

Reply:
> I will change the resample method in the IndividualTimeSeriesResampler to return None if its ring buffer is empty 

Yeah, good point, the return of `resample()` should be `Optional`.

> I'd suggest raising an exception in this case, and specifying in docs when the exception is raised.

Yes, I suggested an `assert`, which raises an exception (`AssertError`). IMHO an `assert` should be enough here as it is a pre-condition of the function, if it fails is because there is a programmer error (whom failed to check if the list was empty before calling the function), other exceptions should be used to signal errors that can't be prevented at programming-time (runtime errors basically).

But if you think we should raise a `ValueError` instead I won't oppose (too strongly :D).

Reply:
So the current state is that the `resample()` method returns `Optional[float]` and the resampling function may be passed an empty list. 

Where should this exception be raised / assertion to take place? The `ResamplingFunction` is provided by the caller so we don't have control over it here. 
```
class TimeSeriesResampler:
    def __init__(
        self,
        resampling_period_s: float,
        max_data_age_in_periods: float,
        resampling_function: ResamplingFunction, <--- provided by the caller
    ) -> None:
        self._resampling_period_s = resampling_period_s
        self._max_data_age_in_periods: float = max_data_age_in_periods
        self._buffer: Deque[Sample] = Deque[Sample]()
        self._resampling_function: ResamplingFunction = resampling_function

    def resample(self) -> Optional[float]:
        now = datetime.now(tz=pytz.UTC)
        threshold = now - timedelta(
            seconds=self._max_data_age_in_periods * self._resampling_period_s
        )
        self.remove_outdated_samples(threshold=threshold)
        # if len(self._buffer) == 0: raise X
        return self._resampling_function(self._buffer, self._resampling_period_s)
```

It would have to be raised before calling the resampling function when the buffer is empty.

Reply:
I don't get this. Why not this?

```py
        self.remove_outdated_samples(threshold=threshold)
        if len(self._buffer) == 0:
            return None
        return self._resampling_function(self._buffer, self._resampling_period_s)
```

Reply:
Yeah, that was my initial idea, but I somehow didn't get it where you and Tiyash wanted to raise an exception?

Reply:
I think the only question was if **inside the resampling function** we raise a `ValueError` exception or just `assert`. Let's go for the assert for now. But the resampler should just not called the resampling function if there is no data to resample.